pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
+	if len(clientConfigMap) <= 0 {
+		delete(objects, object.key)
+		return
+	}
+
+	if certificateChain.CertificatesIssued == nil {
+		certificateChain.CertificatesIssued = map[string]*chain.CertificateIssue{}
+	}
+
+	for name, config := range clientConfigMap {
+		serviceName := config.Service.Name
+		serviceNamespace := config.Service.Namespace
+		serviceHostname := serviceHostname(serviceName, serviceNamespace)
+
+		if _, found := certificateChain.CertificatesIssued[serviceHostname]; !found {
+			certificateChain.CertificatesIssued[serviceHostname] = newCertificateIssue(serviceName, serviceNamespace)
 		}
 
-		for _, clientConfig := range m.clientConfigList(webhook) {
-			// Update the CA bundle at webhook
-			updatedCABundle, err := updateCABundle(clientConfig.CABundle)
-			if err != nil {
-				return errors.Wrap(err, "failed updating CA bundle")
-			}
-			clientConfig.CABundle = updatedCABundle
+		caBundleName := caBundleName(object.key.String(), name)
+		certificateChain.CertificatesIssued[serviceHostname].CACertPEM[caBundleName] = config.CABundle
+		key := newObjectKey("Secret", serviceNamespace, serviceName)
+		if _, found := objects[key]; !found {
+			objects[key] = &keyedObject{key, nil}
 		}
+	}
+}
 
-		err = m.client.Update(context.TODO(), webhook)
-		if err != nil {
-			return err
+// mapWebhookToChain maps a webhook object from certificate chain data.
+func mapWebhookFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	clientConfigList := clientConfigMap(object.kobject)
+	for name, config := range clientConfigList {
+		serviceHostname := serviceHostname(config.Service.Name, config.Service.Namespace)
+		certificateIssue := certificateChain.CertificatesIssued[serviceHostname]
+		if certificateIssue == nil {
+			continue
 		}
-		return nil
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to update webhook CABundle")
+		caBundleName := caBundleName(object.key.String(), name)
+		caBundle := certificateIssue.CACertPEM[caBundleName]
+		if caBundle == nil {
+			continue
+		}
+		config.CABundle = caBundle
+	}
+}
+
+// mapWebhookToChain maps a secret object to certificate chain data.
+func mapSecretToChain(object *keyedObject, objects objectMap, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretToChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretToChain(object, certificateChain)
+}
+
+// mapWebhookToChain maps a secret object from certificate chain data.
+func mapSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretFromChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretFromChain(object, certificateChain)
+}
+
+func mapServiceSecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
```

Comment:
We are exposing ourselves to a panic, I would prefer to return and error and do a controlled downcasting

```golang
secret, ok := object.kobject.(*corev1.Secret)
if !ok {
    return ....
}

For this we have to be able to return an error here.
```